src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/MvcServicesProvider.cs

```diff
@@ -0,0 +1,115 @@
+
+            Host = serviceProvider.GetRequiredService<IMvcRazorHost>();
+            Compiler = serviceProvider.GetRequiredService<CSharpCompiler>();
+
+            FileProvider = serviceProvider.GetRequiredService<IRazorViewEngineFileProviderAccessor>().FileProvider;
+        }
+
+        public IMvcRazorHost Host { get; }
+
+        public CSharpCompiler Compiler { get; }
+
+        public IFileProvider FileProvider { get; }
+
+        private IMvcBuilderConfiguration GetConfigureCompilationAction(string configureCompilationType)
+        {
+            Type type;
+            if (!string.IsNullOrEmpty(configureCompilationType))
+            {
+                type = Type.GetType(configureCompilationType);
+                if (type == null)
+                {
+                    throw new InvalidOperationException($"Unable to find type '{type}.");
+                }
+            }
+            else
+            {
+                var assemblyName = new AssemblyName(_applicationName);
+                var assembly = Assembly.Load(assemblyName);
+                type = assembly
+                    .GetExportedTypes()
+                    .FirstOrDefault(typeof(IMvcBuilderConfiguration).IsAssignableFrom);
+            }
+
+            if (type == null)
+            {
+                return null;
+            }
+
+            var instance = Activator.CreateInstance(type) as IMvcBuilderConfiguration;
+            if (instance == null)
+            {
+                throw new InvalidOperationException($"Type {configureCompilationType} does not implement " +
+                    $"{typeof(IMvcBuilderConfiguration)}.");
+            }
+
+            return instance;
+        }
+
+        private IServiceProvider GetProvider(IMvcBuilderConfiguration mvcBuilderConfiguration)
+        {
+            var services = new ServiceCollection();
+
+            var hostingEnvironment = new HostingEnvironment
+            {
+                ApplicationName = _applicationName,
+                WebRootFileProvider = new PhysicalFileProvider(_projectPath),
+                ContentRootFileProvider = new PhysicalFileProvider(_contentRoot),
+                ContentRootPath = _contentRoot,
+            };
+            var diagnosticSource = new DiagnosticListener("Microsoft.AspNetCore");
+
+            services
+                .AddSingleton<IHostingEnvironment>(hostingEnvironment)
+                .AddSingleton<DiagnosticSource>(diagnosticSource)
+                .AddLogging()
+                .AddSingleton<ObjectPoolProvider, DefaultObjectPoolProvider>();
+
+            var mvcCoreBuilder = services
+                .AddMvcCore()
+                .AddRazorViewEngine();
+
+            var mvcBuilder = new MvcBuilder(mvcCoreBuilder.Services, mvcCoreBuilder.PartManager);
+            mvcBuilderConfiguration?.ConfigureMvc(mvcBuilder);
+
+            return mvcBuilder.Services.BuildServiceProvider();
```

Comment:
/cc @davidfowl - this is what we discussed a while ago about our `AddXyZ(...)` methods not being self-sufficient. This is necessary every where we need to use MVC with DI, but not with hosting.